src/localization.rs

```diff
@@ -0,0 +1,322 @@
+/// about which variant that will be.
+/// In addition to the colon-separated format, using a list with one language per element is
+/// also supported.
+///
+/// Returns the (possibly empty) preference list of languages.
+fn get_language_preferences_from_env(vars: &EnvStack) -> Vec<String> {
+    use crate::wchar::L;
+
+    fn normalize_locale_name(locale: &str) -> String {
+        // Strips off the encoding and modifier parts.
+        let mut normalized_name = String::new();
+        // Strip off encoding and modifier. (We always expect UTF-8 and don't support modifiers.)
+        for c in locale.chars() {
+            if c.is_alphabetic() || c == '_' {
+                normalized_name.push(c);
+            } else {
+                break;
+            }
+        }
+        // At this point, the normalized_name should have the shape `ll` or `ll_CC`.
+        normalized_name
+    }
+
+    fn check_language_var(vars: &EnvStack) -> Option<Vec<String>> {
+        let langs = vars.get(L!("LANGUAGE"))?;
+        let langs = langs.as_list();
+        let filtered_langs: Vec<String> = langs
+            .iter()
+            .filter(|lang| !lang.is_empty())
+            .map(|lang| normalize_locale_name(&lang.to_string()))
+            .collect();
+        if filtered_langs.is_empty() {
+            return None;
+        }
+        Some(filtered_langs)
+    }
+
+    // Locale value is determined by the first of these three variables set to a non-zero
+    // value.
+    if let Some(locale) = vars
+        .get(L!("LC_ALL"))
+        .or_else(|| vars.get(L!("LC_MESSAGES")).or_else(|| vars.get(L!("LANG"))))
+    {
+        let locale = locale.as_string().to_string();
+        if locale.starts_with('C') {
+            // Do not localize in C locale.
+            return vec![];
+        }
+        // `LANGUAGE` has higher precedence than the locale value.
+        if let Some(precedence_list) = check_language_var(vars) {
+            return precedence_list;
+        }
+        // Use the locale value if `LANGUAGE` is not set.
+        vec![normalize_locale_name(&locale)]
+    } else if let Some(precedence_list) = check_language_var(vars) {
+        // Use the `LANGUAGE` value if locale is not set.
+        return precedence_list;
+    } else {
+        // None of the relevant variables are set, so we will not localize.
+        vec![]
+    }
+}
+
+/// Call this when one of `LANGUAGE`, `LC_ALL`, `LC_MESSAGES`, `LANG` changes.
+/// Updates internal state such that the correct localizations will be used in subsequent
+/// localization requests.
+pub fn set_language_precedence_from_env(vars: &EnvStack) {
+    let language_precedence = get_language_preferences_from_env(vars);
+    let languages = language_precedence
+        .iter()
+        .map(|lang| lang.as_str())
+        .collect::<Vec<&str>>();
+    fish_fluent::set_language_precedence(&languages);
+    #[cfg(feature = "localize-messages")]
```

Comment:
(BTW if we write `feature = "localize-messages"` a lot, we can define a cfg for that.)